{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "36fbb8c9",
   "metadata": {},
   "source": [
    "### 1. Install dependencies and load packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "077031ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# or create a conda env\n",
    "!pip install scanpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "804ef016",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import logging\n",
    "\n",
    "# Init logger\n",
    "logger = logging.Logger('logger')\n",
    "logging.basicConfig(\n",
    "    level=logging.INFO,\n",
    "    format='%(asctime)s - %(levelname)s - %(message)s'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ec04869",
   "metadata": {},
   "source": [
    "### 2. Download dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "773a12e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs('data', exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "08232d24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-05-14 17:04:27--  https://zenodo.org/record/7041849/files/ReplogleWeissman2022_rpe1.h5ad\n",
      "Resolving zenodo.org (zenodo.org)... 188.185.45.92, 188.185.43.25, 188.185.48.194, ...\n",
      "Connecting to zenodo.org (zenodo.org)|188.185.45.92|:443... connected.\n",
      "HTTP request sent, awaiting response... 301 MOVED PERMANENTLY\n",
      "Location: /records/7041849/files/ReplogleWeissman2022_rpe1.h5ad [following]\n",
      "--2025-05-14 17:04:27--  https://zenodo.org/records/7041849/files/ReplogleWeissman2022_rpe1.h5ad\n",
      "Reusing existing connection to zenodo.org:443.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1236886900 (1.2G) [application/octet-stream]\n",
      "Saving to: ‘data/ReplogleWeissmann2022_rpe1.h5ad’\n",
      "\n",
      "data/ReplogleWeissm 100%[===================>]   1.15G  29.7MB/s    in 42s     \n",
      "\n",
      "2025-05-14 17:05:09 (28.0 MB/s) - ‘data/ReplogleWeissmann2022_rpe1.h5ad’ saved [1236886900/1236886900]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Bigger dataset (~1.2GB) (one of the ones we mainly use)\n",
    "!wget -O data/ReplogleWeissmann2022_rpe1.h5ad \"https://zenodo.org/record/7041849/files/ReplogleWeissman2022_rpe1.h5ad\"\n",
    "# Smaller dataset (~600MB)\n",
    "!wget -O data/NormanWeissman2019_filtered.h5ad \"https://zenodo.org/record/7041849/files/NormanWeissman2019_filtered.h5ad\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ac83a2c",
   "metadata": {},
   "source": [
    "### 3. Read dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b963976c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/xlv0877/.conda/envs/pydl/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/xlv0877/.conda/envs/pydl/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/xlv0877/.conda/envs/pydl/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/xlv0877/.conda/envs/pydl/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/xlv0877/.conda/envs/pydl/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/xlv0877/.conda/envs/pydl/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/xlv0877/.conda/envs/pydl/lib/python3.11/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "adata_p = 'data/ReplogleWeissmann2022_rpe1.h5ad'\n",
    "adata = sc.read(adata_p)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72757330",
   "metadata": {},
   "source": [
    "### Data uses this format (anndata): https://anndata.readthedocs.io/en/stable/\n",
    "#### Pytorch integration module (scvi-tools): https://scvi-tools.org/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6a87c766",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 247914 × 8749\n",
       "    obs: 'batch', 'gene', 'gene_id', 'transcript', 'gene_transcript', 'guide_id', 'percent_mito', 'UMI_count', 'z_gemgroup_UMI', 'core_scale_factor', 'core_adjusted_UMI_count', 'disease', 'cancer', 'cell_line', 'sex', 'age', 'perturbation', 'organism', 'perturbation_type', 'tissue_type', 'ncounts', 'ngenes', 'nperts', 'percent_ribo', 'celltype'\n",
       "    var: 'chr', 'start', 'end', 'class', 'strand', 'length', 'in_matrix', 'mean', 'std', 'cv', 'fano', 'ensembl_id', 'ncounts', 'ncells'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Print out data overview\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37866030",
   "metadata": {},
   "source": [
    "### adata.shape = (cells x genes)\n",
    "#### - this dataset has 247914 cells and 8749 genes\n",
    "### .obs = observations (pd.DataFrame) --> meta information on each cell\n",
    "#### - obs:  column names\n",
    "### .var = variables (pd.DataFrame) --> meta information on each gene\n",
    "#### - var: column names\n",
    "### adata.X = Expression matrix (CSR/CSC sparse matrix or np.ndarray)\n",
    "#### - this is where the actual gene expression data is stored\n",
    "### adata.layers = Storage for alterations of X, like normalized or scaled expression"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "164a09ac",
   "metadata": {},
   "source": [
    "### 4. Check meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "954c5863",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-05-14 17:20:59,548 - INFO - Dataset has 2394 classes.\n"
     ]
    }
   ],
   "source": [
    "# We are trying to classify the .obs['perturbation'] column with respect to cell type and perturbation type (CRISPR type)\n",
    "\n",
    "# define all labels to classify on\n",
    "cls_labels = ['celltype', 'perturbation_type', 'perturbation']\n",
    "# create classification label\n",
    "adata.obs['cls_label'] = adata.obs[cls_labels].agg(';'.join, axis=1)\n",
    "logging.info(f'Dataset has {adata.obs.cls_label.nunique()} classes.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0584b923",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cls_label\n",
       "retinal pigment epithelial cells;CRISPR;control    11485\n",
       "retinal pigment epithelial cells;CRISPR;TFAM        3580\n",
       "retinal pigment epithelial cells;CRISPR;SLC1A5      1962\n",
       "retinal pigment epithelial cells;CRISPR;GFM1        1699\n",
       "retinal pigment epithelial cells;CRISPR;MRPL36      1686\n",
       "                                                   ...  \n",
       "retinal pigment epithelial cells;CRISPR;ZC3H8          4\n",
       "retinal pigment epithelial cells;CRISPR;CDK7           4\n",
       "retinal pigment epithelial cells;CRISPR;NAT10          3\n",
       "retinal pigment epithelial cells;CRISPR;NPAT           3\n",
       "retinal pigment epithelial cells;CRISPR;NUP93          2\n",
       "Name: count, Length: 2394, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for class imbalances etc.\n",
    "adata.obs.cls_label.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3efff9d2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pydl",
   "language": "python",
   "name": "pydl"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
