<tbody><tr data-index="0" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/AdamsonWeissman2016_GSM2406675_10X001.h5ad" target="_blank">Download</a></td><td>AdamsonWeissman2016</td><td>GSM2406675_10X001</td><td><a href="https://doi.org/10.1016/j.cell.2016.11.048" target="_blank">A multiplexed single-cell CRISPR screening platform enables systematic dissection of the unfolded protein response</a></td><td>Britt Adamson</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>Perturb-seq</td><td>9</td><td>CRISPRi</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="1" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/AdamsonWeissman2016_GSM2406677_10X005.h5ad" target="_blank">Download</a></td><td>AdamsonWeissman2016</td><td>GSM2406677_10X005</td><td><a href="https://doi.org/10.1016/j.cell.2016.11.048" target="_blank">A multiplexed single-cell CRISPR screening platform enables systematic dissection of the unfolded protein response</a></td><td>Britt Adamson</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>Perturb-seq</td><td>20</td><td>CRISPRi</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="2" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/AdamsonWeissman2016_ GSM2406681_10X010.h5ad" target="_blank">Download</a></td><td>AdamsonWeissman2016</td><td>GSM2406681_10X010</td><td><a href="https://doi.org/10.1016/j.cell.2016.11.048" target="_blank">A multiplexed single-cell CRISPR screening platform enables systematic dissection of the unfolded protein response</a></td><td>Britt Adamson</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>Perturb-seq</td><td>114</td><td>CRISPRi</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="3" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/AissaBenevolenskaya2021.h5ad" target="_blank">Download</a></td><td>AissaBenevolenskaya2021</td><td></td><td><a href="https://doi.org/10.1038/s41467-021-21884-z" target="_blank">Single-cell transcriptional changes associated with
drug tolerance and response to combination
therapies in cancer</a></td><td>Alexandre Aissa</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>Drop-Seq</td><td>4</td><td>drugs</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="4" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/ChangYe2021.h5ad" target="_blank">Download</a></td><td>ChangYe2021</td><td></td><td><a href="https://doi.org/10.1038/s41587-021-01005-3" target="_blank">Identifying transcriptional programs underlying cancer drug response with TraCe-seq</a></td><td>Matthew Chang</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>clonal tagging (TraCe-seq)</td><td>4</td><td>drugs</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="5" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/DatlingerBock2017.h5ad" target="_blank">Download</a></td><td>DatlingerBock2017</td><td></td><td><a href="https://doi.org/10.1038/nmeth.4177" target="_blank">Pooled CRISPR screening with single-cell transcriptome readout</a></td><td>Paul Datlinger</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>CROP-seq</td><td>97</td><td>CRISPR-cas9, TCR stim</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="6" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/DatlingerBock2021.h5ad" target="_blank">Download</a></td><td>DatlingerBock2021</td><td></td><td><a href="https://doi.org/10.1038/s41592-021-01153-z" target="_blank">Ultra-high-throughput single-cell RNA sequencing and perturbation screening with combinatorial fluidic indexing</a></td><td>Paul Datlinger</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>scifi-RNA-seq (derived from CROP-seq)</td><td>48</td><td>CRISPR-cas9, TCR stim</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="7" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/DixitRegev2016.h5ad" target="_blank">Download</a></td><td>DixitRegev2016</td><td></td><td><a href="https://doi.org/10.1016/j.cell.2016.11.038" target="_blank">Perturb-Seq: Dissecting Molecular Circuits with Scalable Single-Cell RNA Profiling of Pooled Genetic Screens</a></td><td>Atray Dixit</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>Perturb-seq</td><td>31</td><td>CRISPR-cas9</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="8" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/FrangiehIzar2021_RNA.h5ad" target="_blank">Download</a></td><td>FrangiehIzar2021</td><td>RNA</td><td><a href="https://doi.org/10.1038/s41588-021-00779-1" target="_blank">Multimodal pooled Perturb-CITE-seq screens in patient models define mechanisms of cancer immune evasion</a></td><td>Chris Frangieh</td><td style="background-color: rgb(31, 119, 180) !important;">RNA + protein (RNA)</td><td>Perturb-CITE-seq</td><td>249</td><td>CRISPR-cas9</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="9" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/FrangiehIzar2021_protein.h5ad" target="_blank">Download</a></td><td>FrangiehIzar2021</td><td>protein</td><td><a href="https://doi.org/10.1038/s41588-021-00779-1" target="_blank">Multimodal pooled Perturb-CITE-seq screens in patient models define mechanisms of cancer immune evasion</a></td><td>Chris Frangieh</td><td style="background-color: rgb(214, 39, 40) !important;">RNA + protein (protein)</td><td>Perturb-CITE-seq</td><td>249</td><td>CRISPR-cas9</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="10" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/GasperiniShendure2019_atscale.h5ad" target="_blank">Download</a></td><td>GasperiniShendure2019</td><td>atscale</td><td><a href="https://doi.org/10.1016/j.cell.2018.11.029" target="_blank">A Genome-wide Framework for Mapping Gene Regulation via Cellular Genetic Screens</a></td><td>Molly Gasperini</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>CROP-seq</td><td>43314<a tabindex="0" role="button" href="#" data-toggle="popover" data-trigger="focus" data-html="true" data-content="<div style=&quot;overflow: auto; max-height: 30vh; max-width: 25vw;&quot;>43314 (includes combinations)</div>" data-original-title="" title="">...</a></td><td>CRISPRi</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="11" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/GasperiniShendure2019_highMOI.h5ad" target="_blank">Download</a></td><td>GasperiniShendure2019</td><td>highMOI</td><td><a href="https://doi.org/10.1016/j.cell.2018.11.029" target="_blank">A Genome-wide Framework for Mapping Gene Regulation via Cellular Genetic Screens</a></td><td>Molly Gasperini</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>CROP-seq</td><td>39087<a tabindex="0" role="button" href="#" data-toggle="popover" data-trigger="focus" data-html="true" data-content="<div style=&quot;overflow: auto; max-height: 30vh; max-width: 25vw;&quot;>39087 (includes combinations)</div>" data-original-title="" title="">...</a></td><td>CRISPRi</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="12" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/GasperiniShendure2019_lowMOI.h5ad" target="_blank">Download</a></td><td>GasperiniShendure2019</td><td>lowMOI</td><td><a href="https://doi.org/10.1016/j.cell.2018.11.029" target="_blank">A Genome-wide Framework for Mapping Gene Regulation via Cellular Genetic Screens</a></td><td>Molly Gasperini</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>CROP-seq</td><td>16531<a tabindex="0" role="button" href="#" data-toggle="popover" data-trigger="focus" data-html="true" data-content="<div style=&quot;overflow: auto; max-height: 30vh; max-width: 25vw;&quot;>16531 (includes combinations)</div>" data-original-title="" title="">...</a></td><td>CRISPRi</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="13" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/GehringPachter2019.h5ad" target="_blank">Download</a></td><td>GehringPachter2019</td><td></td><td><a href="https://doi.org/10.1038/s41587-019-0372-z" target="_blank">Highly multiplexed single-cell RNA-seq by DNA oligonucleotide tagging of cellular proteins</a></td><td>Jase Gehring</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>96-plex scRNA-seq
oligo barcodes</td><td>4</td><td>drugs</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="14" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7058382/files/Liscovitch-BrauerSanjana2021_K562_1.zip" target="_blank">Download</a></td><td>Liscovitch-BrauerSanjana2021</td><td>K562_1</td><td><a href="https://doi.org/10.1038/s41587-021-00902-x" target="_blank">Profiling the genetic determinants of chromatin accessibility with scalable single-cell CRISPR screens
 </a></td><td>Liscovitch-Brauer, N.</td><td style="background-color: rgb(255, 127, 14) !important;">ATAC</td><td>CRISPR–sciATAC</td><td>22</td><td>CRISPR-cas9</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="15" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7058382/files/Liscovitch-BrauerSanjana2021_K562_2.zip" target="_blank">Download</a></td><td>Liscovitch-BrauerSanjana2021</td><td>K562_2</td><td><a href="https://doi.org/10.1038/s41587-021-00902-x" target="_blank">Profiling the genetic determinants of chromatin accessibility with scalable single-cell CRISPR screens
 </a></td><td>Liscovitch-Brauer, N.</td><td style="background-color: rgb(255, 127, 14) !important;">ATAC</td><td>CRISPR–sciATAC</td><td>84</td><td>CRISPR-cas9</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="16" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/McFarlandTsherniak2020.h5ad" target="_blank">Download</a></td><td>McFarlandTsherniak2020</td><td></td><td><a href="https://doi.org/10.1038/s41467-020-17440-w" target="_blank">Multiplexed single-cell transcriptional response profiling to define cancer vulnerabilities and therapeutic mechanism of action</a></td><td>James McFarland</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>MIX-seq (pooled scRNA-seq)</td><td>18</td><td>drugs, CRISPR-cas9</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="17" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7058382/files/MimitouSmibert2021.zip" target="_blank">Download</a></td><td>MimitouSmibert2021</td><td></td><td><a href="https://doi.org/10.1038/s41587-021-00927-2" target="_blank">Scalable, multimodal profiling of chromatin accessibility, gene expression and protein levels in single cells</a></td><td>Eleni P. Mimitou</td><td style="background-color: rgb(188, 189, 34) !important;">ATAC + protein</td><td>ASAP-seq</td><td>6</td><td>CRISPR-cas9</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="18" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/NormanWeissman2019_filtered.h5ad" target="_blank">Download</a></td><td>NormanWeissman2019</td><td>filtered</td><td><a href="https://doi.org/10.1126/science.aax4438" target="_blank">Exploring genetic interaction manifolds constructed from rich single-cell phenotypes</a></td><td>Thomas Norman</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>Perturb-seq</td><td>237</td><td>CRISPRa</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="19" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/PapalexiSatija2021_eccite_arrayed_protein.h5ad" target="_blank">Download</a></td><td>PapalexiSatija2021</td><td>eccite_arrayed_protein</td><td><a href="https://doi.org/10.1038/s41588-021-00778-2" target="_blank">Characterizing the molecular regulation of inhibitory immune checkpoints with multimodal single-cell screens</a></td><td>Papalexi, E.</td><td style="background-color: rgb(214, 39, 40) !important;">RNA + protein (protein)</td><td>ECCITE-seq</td><td>11</td><td>CRISPR-cas9</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="20" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/PapalexiSatija2021_eccite_arrayed_RNA.h5ad" target="_blank">Download</a></td><td>PapalexiSatija2021</td><td>eccite_arrayed_RNA</td><td><a href="https://doi.org/10.1038/s41588-021-00778-2" target="_blank">Characterizing the molecular regulation of inhibitory immune checkpoints with multimodal single-cell screens</a></td><td>Papalexi, E.</td><td style="background-color: rgb(31, 119, 180) !important;">RNA + protein (RNA)</td><td>ECCITE-seq</td><td>11</td><td>CRISPR-cas9</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="21" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/PapalexiSatija2021_eccite_protein.h5ad" target="_blank">Download</a></td><td>PapalexiSatija2021</td><td>eccite_protein</td><td><a href="https://doi.org/10.1038/s41588-021-00778-2" target="_blank">Characterizing the molecular regulation of inhibitory immune checkpoints with multimodal single-cell screens</a></td><td>Papalexi, E.</td><td style="background-color: rgb(214, 39, 40) !important;">RNA + protein (protein)</td><td>ECCITE-seq</td><td>99</td><td>CRISPR-cas9</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="22" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/PapalexiSatija2021_eccite_RNA.h5ad" target="_blank">Download</a></td><td>PapalexiSatija2021</td><td>eccite_RNA</td><td><a href="https://doi.org/10.1038/s41588-021-00778-2" target="_blank">Characterizing the molecular regulation of inhibitory immune checkpoints with multimodal single-cell screens</a></td><td>Papalexi, E.</td><td style="background-color: rgb(31, 119, 180) !important;">RNA + protein (RNA)</td><td>ECCITE-seq</td><td>99</td><td>CRISPR-cas9</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="23" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7058382/files/PierceGreenleaf2021_K562.zip" target="_blank">Download</a></td><td>PierceGreenleaf2021</td><td>K562</td><td><a href="https://doi.org/10.1038/s41467-021-23213-w" target="_blank">High-throughput single-cell chromatin accessibility CRISPR screens enable unbiased identification of regulatory networks in cancer
</a></td><td>Sarah Pierce</td><td style="background-color: rgb(255, 127, 14) !important;">ATAC</td><td>Spear-ATAC</td><td>41</td><td>CRISPRi</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="24" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7058382/files/PierceGreenleaf2021_MCF7.zip" target="_blank">Download</a></td><td>PierceGreenleaf2021</td><td>MCF7</td><td><a href="https://doi.org/10.1038/s41467-021-23213-w" target="_blank">High-throughput single-cell chromatin accessibility CRISPR screens enable unbiased identification of regulatory networks in cancer
</a></td><td>Sarah Pierce</td><td style="background-color: rgb(255, 127, 14) !important;">ATAC</td><td>Spear-ATAC</td><td>41</td><td>CRISPRi</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="25" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7058382/files/PierceGreenleaf2021_GM12878.zip" target="_blank">Download</a></td><td>PierceGreenleaf2021</td><td>GM12878</td><td><a href="https://doi.org/10.1038/s41467-021-23213-w" target="_blank">High-throughput single-cell chromatin accessibility CRISPR screens enable unbiased identification of regulatory networks in cancer
</a></td><td>Sarah Pierce</td><td style="background-color: rgb(255, 127, 14) !important;">ATAC</td><td>Spear-ATAC</td><td>41</td><td>CRISPRi</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="26" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/ReplogleWeissman2022_K562_essential.h5ad" target="_blank">Download</a></td><td>ReplogleWeissman2022</td><td>K562_essential</td><td><a href="https://doi.org/10.1016/j.cell.2022.05.013" target="_blank">Mapping information-rich genotype-phenotype landscapes with genome-scale Perturb-seq</a></td><td>Joseph Replogle</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>Perturb-seq</td><td>2058</td><td>CRISPRi</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="27" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/ReplogleWeissman2022_K562_gwps.h5ad" target="_blank">Download</a></td><td>ReplogleWeissman2022</td><td>K562_gwps</td><td><a href="https://doi.org/10.1016/j.cell.2022.05.013" target="_blank">Mapping information-rich genotype-phenotype landscapes with genome-scale Perturb-seq</a></td><td>Joseph Replogle</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>Perturb-seq</td><td>9867</td><td>CRISPRi</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="28" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/ReplogleWeissman2022_rpe1.h5ad" target="_blank">Download</a></td><td>ReplogleWeissman2022</td><td>rpe1</td><td><a href="https://doi.org/10.1016/j.cell.2022.05.013" target="_blank">Mapping information-rich genotype-phenotype landscapes with genome-scale Perturb-seq</a></td><td>Joseph Replogle</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>Perturb-seq</td><td>2394</td><td>CRISPRi</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="29" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/SchiebingerLander2019_GSE106340.h5ad" target="_blank">Download</a></td><td>SchiebingerLander2019</td><td>GSE106340</td><td><a href="https://doi.org/10.1016/j.cell.2019.01.006" target="_blank">Optimal-Transport Analysis of Single-Cell Gene Expression Identifies Developmental Trajectories in Reprogramming</a></td><td>Geoffrey Schiebinger</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>plain scRNA-seq</td><td>2</td><td>cytokines</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="30" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/SchiebingerLander2019_GSE115943.h5ad" target="_blank">Download</a></td><td>SchiebingerLander2019</td><td>GSE115943</td><td><a href="https://doi.org/10.1016/j.cell.2019.01.006" target="_blank">Optimal-Transport Analysis of Single-Cell Gene Expression Identifies Developmental Trajectories in Reprogramming</a></td><td>Geoffrey Schiebinger</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>plain scRNA-seq</td><td>3</td><td>cytokines</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="31" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#"><i class="fa fa-plus"></i></a>
        </td><td><a href="https://zenodo.org/record/7041849/files/SchraivogelSteinmetz2020_TAP_SCREEN__chromosome_11_screen.h5ad" target="_blank">Download</a></td><td>SchraivogelSteinmetz2020</td><td>TAP_SCREEN__chromosome_11_screen</td><td><a href="https://doi.org/10.1038/s41592-020-0837-5" target="_blank">Targeted Perturb-seq enables genome-scale genetic screens in single cells</a></td><td>Daniel Schraivogel</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>TAP-seq</td><td>3105</td><td>CRISPRi</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="32" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#"><i class="fa fa-plus"></i></a>
        </td><td><a href="https://zenodo.org/record/7041849/files/SchraivogelSteinmetz2020_TAP_SCREEN__chromosome_8_screen.h5ad" target="_blank">Download</a></td><td>SchraivogelSteinmetz2020</td><td>TAP_SCREEN__chromosome_8_screen</td><td><a href="https://doi.org/10.1038/s41592-020-0837-5" target="_blank">Targeted Perturb-seq enables genome-scale genetic screens in single cells</a></td><td>Daniel Schraivogel</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>TAP-seq</td><td>4115</td><td>CRISPRi</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="33" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/ShifrutMarson2018.h5ad" target="_blank">Download</a></td><td>ShifrutMarson2018</td><td></td><td><a href="https://doi.org/10.1016/j.cell.2018.10.024" target="_blank">Genome-wide CRISPR Screens in Primary Human T Cells Reveal Key Regulators of Immune Function</a></td><td>Eric Shifrut</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>CROP-seq</td><td>49</td><td>CRISPR-cas9, TCR stim</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="34" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/SrivatsanTrapnell2020_sciplex2.h5ad" target="_blank">Download</a></td><td>SrivatsanTrapnell2020</td><td>sciplex2</td><td><a href="https://doi.org/10.1126/science.aax6234" target="_blank">Massively multiplex chemical transcriptomics at single-cell resolution</a></td><td>Sanjay Srivatsan</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>sci-Plex (nuclear hashing)</td><td>5</td><td>drugs</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="35" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/SrivatsanTrapnell2020_sciplex3.h5ad" target="_blank">Download</a></td><td>SrivatsanTrapnell2020</td><td>sciplex3</td><td><a href="https://doi.org/10.1126/science.aax6234" target="_blank">Massively multiplex chemical transcriptomics at single-cell resolution</a></td><td>Sanjay Srivatsan</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>sci-Plex (nuclear hashing)</td><td>189</td><td>drugs</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="36" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/SrivatsanTrapnell2020_sciplex4.h5ad" target="_blank">Download</a></td><td>SrivatsanTrapnell2020</td><td>sciplex4</td><td><a href="https://doi.org/10.1126/science.aax6234" target="_blank">Massively multiplex chemical transcriptomics at single-cell resolution</a></td><td>Sanjay Srivatsan</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>sci-Plex (nuclear hashing)</td><td>8</td><td>drugs</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="37" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/TianKampmann2019_day7neuron.h5ad" target="_blank">Download</a></td><td>TianKampmann2019</td><td>day7neuron</td><td><a href="https://doi.org/10.1016/j.neuron.2019.07.014" target="_blank">CRISPR Interference-Based Platform for Multimodal Genetic Screens in Human iPSC-Derived Neurons</a></td><td>Ruilin Tian</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>CROP-seq</td><td>27</td><td>CRISPRi</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="38" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/TianKampmann2019_iPSC.h5ad" target="_blank">Download</a></td><td>TianKampmann2019</td><td>iPSC</td><td><a href="https://doi.org/10.1016/j.neuron.2019.07.014" target="_blank">CRISPR Interference-Based Platform for Multimodal Genetic Screens in Human iPSC-Derived Neurons</a></td><td>Ruilin Tian</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>CROP-seq</td><td>27</td><td>CRISPRi</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="39" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/TianKampmann2021_CRISPRa.h5ad" target="_blank">Download</a></td><td>TianKampmann2021</td><td>CRISPRa</td><td><a href="https://doi.org/10.1038/s41593-021-00862-0" target="_blank">Genome-wide CRISPRi/a screens in human neurons link lysosomal failure to ferroptosis</a></td><td>Ruilin Tian</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>CROP-seq</td><td>101</td><td>CRISPRa</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="40" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/TianKampmann2021_CRISPRi.h5ad" target="_blank">Download</a></td><td>TianKampmann2021</td><td>CRISPRi</td><td><a href="https://doi.org/10.1038/s41593-021-00862-0" target="_blank">Genome-wide CRISPRi/a screens in human neurons link lysosomal failure to ferroptosis</a></td><td>Ruilin Tian</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>CROP-seq</td><td>185</td><td>CRISPRi</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="41" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/WeinrebKlein2020.h5ad" target="_blank">Download</a></td><td>WeinrebKlein2020</td><td></td><td><a href="https://doi.org/10.1126/science.aaw3381" target="_blank">Lineage tracing on transcriptional landscapes links state to fate during differentiation</a></td><td>Caleb Weinreb</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>clonal tagging (LARRY)</td><td>5</td><td>cytokines</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="42" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/XieHon2017.h5ad" target="_blank">Download</a></td><td>XieHon2017</td><td></td><td><a href="https://doi.org/10.1016/j.molcel.2017.03.007" target="_blank">Multiplexed engineering and analysis of endogenous enhancer activity in single cells</a></td><td>Shiqi Xie</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>Mosaic-seq</td><td>241</td><td>CRISPR-cas9</td><td style="background-color: rgb(214, 39, 40) !important;">n</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr><tr data-index="43" data-has-detail-view="true"><td>
          <a class="detail-icon" href="#">
          <i class="fa fa-plus"></i>
          </a>
        </td><td><a href="https://zenodo.org/record/7041849/files/ZhaoSims2021.h5ad" target="_blank">Download</a></td><td>ZhaoSims2021</td><td></td><td><a href="https://doi.org/10.1186/s13073-021-00894-y" target="_blank">Deconvolution of Cell Type-Specific Drug Responses in Human Tumor Tissue with Single-Cell RNA-seq</a></td><td>Wenting Zhao</td><td style="background-color: rgb(31, 119, 180) !important;">RNA</td><td>scRNA-seq</td><td>7</td><td>drugs</td><td style="background-color: rgb(44, 160, 44) !important;">y</td><td class="plotcell"><div class="vega-embed" role="graphics-document" aria-roledescription="visualization" aria-label="Vega visualization" style="cursor: default;"><canvas width="61" height="34" class="marks"></canvas><form class="vega-bindings"></form></div></td></tr></tbody>